File: jumpToTheEnd.js
/**
 * BY Facebook / leetoce- 45
 * Starting at index 0, for an element n at index i, you are allowed to jump at most n indexes ahead. Given a list of numbers, find the minimum number of jumps to reach the end of the list.

Example:
Input: [3, 2, 5, 1, 1, 9, 3, 4]
Output: 2
Explanation:

The minimum number of jumps to get to the end of the list is 2: 
3 -> 5 -> 4

*/
// Solution1 : O(n)

var jump = nums => {
    if (nums.length == 1) return 0;
  
    var jumps = 0
    var maxReach = nums[0]
    var steps = nums[0]
  
    for (var i=1; i < nums.length - 1; i++) {
      maxReach = Math.max(maxReach, i + nums[i])
      steps--
      if (steps == 0){
        jumps++
        steps = maxReach - i
      }
    }
    return jumps + 1
  }
  
// Solution 2:
var minJump = nums => {
     //Note, you can assume that you can always reach the last index. count is the total steps.
     var max = nums[0], len = nums.length, startIndex = 0, endIndex = nums[0], count = 1
     if (len < 2) return 0

     //Every time we calculate the range that next jump will result in;
    //max stores the maximum index that next jump can reach, which is the next range's right boundary;
    //Next range's left boundary will be current range's right boundary plus 1; then add 1 step to count;
    //Thus we get the next range. Before loop next range we check if its right boundary is bigger or equal to the largest index in nums (len - 1);
    //If true exit loop and return count (steps);

    while (max < len - 1){
        var n = startIndex, m= endIndex;
        startIndex = endIndex + 1
        for (var i= n; i <= m; i++){
            if(i + nums[i] > max) {
                max = i + nums[i]
                endIndex = max
            }
        }
        count++
    }
     return count
}
  
  nums = [3, 2, 5, 1, 1, 9, 3, 4]
  console.log(minJump(nums));

File: jumpToTheEnd.py
''' BY Facebook Leetcode- 45
Starting at index 0, for an element n at index i, you are allowed to jump at most n indexes ahead. Given a list of numbers, find the minimum number of jumps to reach the end of the list.

Example:
Input: [3, 2, 5, 1, 1, 9, 3, 4]
Output: 2
Explanation:

The minimum number of jumps to get to the end of the list is 2: 
3 -> 5 -> 4
solution:

1. We start travering the array from start
2. While traversing, we keep a track on maximum reachable index and update it accordingly.
3. If we reach the previous reachable index, this implies we have reached this index with 
   current jumps and now we can reach the next maximum possible index by current jumps+1 jumps 
   and update the previous reachable index to maximum reachable index. Now, if updated previous 
   reachable index is greater than equal to last index, get out of loop.

'''

class Solution:
    def jump(self, nums):
        if len(nums) == 1: return 0

        reachableIndex = 0
        previousReachableIndex = 0
        jump = 0

        for curr in range(len(nums)):
            if curr + nums[curr] >= reachableIndex:
                reachableIndex = curr + nums[curr]

            if curr == previousReachableIndex:
                jump += 1
                previousReachableIndex = reachableIndex
                if previousReachableIndex >= len(nums) - 1:
                    break
        return jump

nums = [3, 2, 5, 1, 1, 9, 3, 4]
x = Solution().jump(nums)
print(x)